shell/platform/darwin/ios/framework/Source/FlutterViewController.mm

```diff
@@ -1102,29 +1114,105 @@ - (void)keyboardWillChangeFrame:(NSNotification*)notification {
     }
   }
 
+  // Ignore keyboard notifications if engine’s viewController is not current viewController
+  if ([_engine.get() viewController] != self) {
+    return;
+  }
+
   CGRect keyboardFrame = [[info objectForKey:UIKeyboardFrameEndUserInfoKey] CGRectValue];
   CGRect screenRect = [[UIScreen mainScreen] bounds];
 
+  // Get the animation duration and curve
+  NSTimeInterval duration =
+      [[info objectForKey:UIKeyboardAnimationDurationUserInfoKey] doubleValue];
+  UIViewAnimationCurve curve =
+      (UIViewAnimationCurve)[[info objectForKey:UIKeyboardAnimationCurveUserInfoKey] integerValue];
+
   // Considering the iPad's split keyboard, Flutter needs to check if the keyboard frame is present
   // in the screen to see if the keyboard is visible.
   if (CGRectIntersectsRect(keyboardFrame, screenRect)) {
     CGFloat bottom = CGRectGetHeight(keyboardFrame);
     CGFloat scale = [UIScreen mainScreen].scale;
-
     // The keyboard is treated as an inset since we want to effectively reduce the window size by
     // the keyboard height. The Dart side will compute a value accounting for the keyboard-consuming
     // bottom padding.
-    _viewportMetrics.physical_view_inset_bottom = bottom * scale;
+    self.targetViewInsetBottom = bottom * scale;
   } else {
-    _viewportMetrics.physical_view_inset_bottom = 0;
+    self.targetViewInsetBottom = 0;
   }
-
-  [self updateViewportMetrics];
+  [self startKeyBoardAnimation:duration curve:curve];
 }
 
 - (void)keyboardWillBeHidden:(NSNotification*)notification {
-  _viewportMetrics.physical_view_inset_bottom = 0;
-  [self updateViewportMetrics];
+  // When keyboard hide,the keyboardWillChangeFrame function will be called to update viewport
+  // metrics. So do not call [self updateViewportMetrics] here again
+}
+
+- (void)startKeyBoardAnimation:(NSTimeInterval)duration curve:(UIViewAnimationCurve)curve {
+  // If current physical_view_inset_bottom == targetViewInsetBottom,do nothing
+  if (_viewportMetrics.physical_view_inset_bottom == self.targetViewInsetBottom) {
+    return;
+  }
+
+  // When call this method first time,
+  // initialize the keyboardAnimationView to get animation interpolation during animation
+  if ([self keyboardAnimationView] == nil) {
+    _keyboardAnimationView.reset([[UIView alloc] init]);
```

Comment:
This view could interfere with touch events.  We should make sure that there is no possible way that it could get touch events: https://developer.apple.com/documentation/uikit/uiview/1622577-userinteractionenabled?language=objc

Reply:
done